### PR TITLE
[SYCL][AOT] Enhance opencl-aot to compile or link program

### PR DESCRIPTION
**Pull request by linkren**
This patch implements 3 invocation types:

* opencl-aot -device=<value> -cmd=link -spv=<spirv_input>
             -ir=<output_aocx>
  This compiles a SPIR-V program to an aocx (i.e. the standard ELF file
  from opencl_rt compiles)

* opencl-aot -device=<value> -cmd=link
             -binary=<comma_separated_list_of_binary_filename>
             -ir=<output_aocx>
  This links LLVM IR and object files to generate an aocx

* opencl-aot -device=<value> -cmd=compile -input=<opencl_input_file>
             -ir=<output_aocx>
  This compiles OpenCL source to an aocx

### CODE REVIEW

File: opencl-aot/source/main.cpp [hunk 1]
```diff
@@ -264,6 +282,23 @@ int main(int Argc, char *Argv[]) {
...
+          clEnumVal(
+              build,
+              "Build executable IR from OpenCL source or OpenCL .obj file"),
+          clEnumVal(compile, "Compile from OpenCL source to .obj"),
+          clEnumVal(link, "Link OpenCL .obj file(s)")));
+  cl::opt<std::string> OptInput("input", cl::desc("Input OpenCL C file"));
+  cl::opt<std::string> OptIR("ir", cl::desc("Output file"));
+  cl::opt<std::string> OptSPIRV("spv", cl::desc("Input SPIR-V file"));
+  cl::opt<std::string> OptBinaries("binary",
+                                   cl::desc("Input OpenCL .obj files to link"));
```
Comment (AlexeySachkov) [inline opencl-aot/source/main.cpp:300]: As we discussed offline, these options are only needed for compatibility with some other Intel-provided tool.
Please put a comment describing that and add `TODO`/`FIXME` to delete this option once transition to a regular `-o` and regular input argument is complete.

Moreover, I suggest that for inputs consistency checking we use actual filetypes which we detect by looking into file content like it is already done:
https://github.com/intel/llvm/blob/41443127d775c6e09686dd11dbffcf9186c44e01/opencl-aot/source/main.cpp#L146-L161

And I also suggest to use [aliases](https://llvm.org/docs/CommandLine.html#argument-aliases) to leave or logic to work with regular input/output arguments and declare new arguments as aliases to the existing ones

File: opencl-aot/source/main.cpp [hunk 2]
```diff
@@ -303,15 +338,100 @@ int main(int Argc, char *Argv[]) {
                                         cl::desc("Set OpenCL build options"),
                                         cl::value_desc("build options"));
 
+  // Adding custom --version implementation.
+  cl::SetVersionPrinter(
+      [](llvm::raw_ostream &OS) { OS << "Intel(R) SDK for OpenCL(TM)\n"; });
```
Comment (AlexeySachkov) [inline opencl-aot/source/main.cpp:343]: I'm not sure that this is a proper version string for an open-source tool: this project is not related to `Intel(R) SDK for OpenCL(TM)` and such customization should be done downstream if needed

File: opencl-aot/source/main.cpp [hunk 3]
```diff
@@ -303,15 +338,100 @@ int main(int Argc, char *Argv[]) {
...
+      }
+
+      break;
+    }
+
+    // Output file name is specified by -ir if --device=fpga_fast_emu.
+    if (!OptIR.getNumOccurrences()) {
+      std::cerr << "-cmd=<value> needs -ir option\n";
+      return OPENCL_AOT_OPTIONS_DEFICIENCY_FAILURE;
+    }
```
Comment (AlexeySachkov) [inline opencl-aot/source/main.cpp:413]: If `-ir` is only introduced for compatibility reasons, I suggest do not introduce this restriction and allow using regular `-o` even for `fpga_fast_emu` device and `-cmd` option, because such restriction basically blocks any transition to a regular `-o`

File: opencl-aot/source/main.cpp [hunk 4]
```diff
@@ -303,15 +338,100 @@ int main(int Argc, char *Argv[]) {
...
+    switch (OptCommand) {
+    case Commands::build:
+      // Ensure we get either -input or -binary
+      if (!OptBinaries.getNumOccurrences() && !OptInput.getNumOccurrences()) {
+        std::cerr << "-cmd=build needs -input or -binary option\n";
+        return OPENCL_AOT_OPTIONS_DEFICIENCY_FAILURE;
+      }
+      // Ensure we don't have both
+      if (OptBinaries.getNumOccurrences() && OptInput.getNumOccurrences()) {
+        std::cerr << "-cmd=link accepts only one of -input or -binary option\n";
```
Comment (AlexeySachkov) [inline opencl-aot/source/main.cpp:367]:
```suggestion
        std::cerr << "-cmd=build accepts only one of -input or -binary option\n";
```

File: opencl-aot/source/main.cpp [hunk 5]
```diff
@@ -303,15 +338,100 @@ int main(int Argc, char *Argv[]) {
...
+    // Output file name is specified by -ir if --device=fpga_fast_emu.
+    if (!OptIR.getNumOccurrences()) {
+      std::cerr << "-cmd=<value> needs -ir option\n";
+      return OPENCL_AOT_OPTIONS_DEFICIENCY_FAILURE;
+    }
+    OutputFileName = OptIR;
+  } else if (OptInput.getNumOccurrences() || OptBinaries.getNumOccurrences() ||
+             OptSPIRV.getNumOccurrences() || OptIR.getNumOccurrences()) {
+    std::cerr << "-cmd option must be specified\n";
+    return OPENCL_AOT_OPTIONS_DEFICIENCY_FAILURE;
```
Comment (AlexeySachkov) [inline opencl-aot/source/main.cpp:418]: Why such restriction is needed? I think that we can always come up with some default actions based on a file type.

Basically for `binary` input you can't do anything except link it, for `OpenCL C` input the best default action would be to perform a full build, for `SPIR-V` it would also `build` or `link` (they are equal here)

File: opencl-aot/source/main.cpp [hunk 6]
```diff
@@ -426,9 +556,30 @@ int main(int Argc, char *Argv[]) {
   }
   std::cout << "Using build options: " << BuildOptions << '\n';
 
-  // step 8: build OpenCL program
-  CLErr = clBuildProgram(ProgramUPtr.get(), 1, &DeviceId, BuildOptions.c_str(),
-                         nullptr, nullptr);
+  // step 8: compile | build | link OpenCL program
+  if (OptCommand.getNumOccurrences() && OptCommand == Commands::compile) {
+    assert(OptDevice == fpga_fast_emu &&
+           "Use -cmd=compile with --device=fpga_fast_emu only");
```
Comment (AlexeySachkov) [inline opencl-aot/source/main.cpp:562]: Why do we have this assertion? `clCompileProgram` is a pretty standard API which must be supported regardless of target device (at least in OpenCL 1.2, 2.*)

File: opencl-aot/source/main.cpp [hunk 7]
```diff
@@ -426,9 +556,30 @@ int main(int Argc, char *Argv[]) {
...
+    CLErr =
+        clCompileProgram(ProgramUPtr.get(), 1, &DeviceId, BuildOptions.c_str(),
+                         0, nullptr, nullptr, nullptr, nullptr);
+  } else if (OptCommand.getNumOccurrences() && OptCommand == Commands::link &&
+             OptBinaries.getNumOccurrences()) {
+    assert(OptDevice == fpga_fast_emu &&
+           "Use -cmd=link with --device=fpga_fast_emu only");
```
Comment (AlexeySachkov) [inline opencl-aot/source/main.cpp:569]: Same question as above about this assertion

**Review by AlexeySachkov — CHANGES_REQUESTED** [review]

File: opencl-aot/source/main.cpp (hunk 2 above)
Comment (linkren) [reply opencl-aot/source/main.cpp]: Removed this customization.

File: opencl-aot/source/main.cpp (hunk 3 above)
Comment (linkren) [reply opencl-aot/source/main.cpp]: Done.

File: opencl-aot/source/main.cpp (hunk 4 above)
Comment (linkren) [reply opencl-aot/source/main.cpp]: Done.

File: opencl-aot/source/main.cpp (hunk 6 above)
Comment (linkren) [reply opencl-aot/source/main.cpp]: Removed.

File: opencl-aot/source/main.cpp (hunk 7 above)
Comment (linkren) [reply opencl-aot/source/main.cpp]: Removed.

File: opencl-aot/source/main.cpp (hunk 5 above)
Comment (linkren) [reply opencl-aot/source/main.cpp]: Removed. However, for back compatibility, we only accept comma-separated list of input files and this only work for -cmd=link.

**Review by AlexeySachkov — APPROVED** [review]
I don't have any further comments/objections, but I would like to hear feedback from @dm-vodopyanov

File: opencl-aot/source/main.cpp [hunk 8]
```diff
@@ -253,17 +265,46 @@ getCompilerBuildLog(const CLProgramUPtr &ProgramUPtr, cl_device_id Device) {
...
           clEnumVal(fpga_fast_emu,
                     "Intel(R) FPGA Emulation Platform for OpenCL(TM) device")));
+
+  enum Commands { build, compile, link };
+  cl::opt<Commands> OptCommand(
+      "cmd", cl::desc("Command"),
+      cl::values(
+          clEnumVal(
+              build,
+              "Build executable IR from OpenCL source or OpenCL .obj file"),
```
Comment (dm-vodopyanov) [inline opencl-aot/source/main.cpp:291]:
```suggestion
              "Build (compile and link) OpenCL program binary from SPIR-V or OpenCL program binary or OpenCL C source."),
```

File: opencl-aot/source/main.cpp [hunk 9]
```diff
@@ -253,17 +265,46 @@ getCompilerBuildLog(const CLProgramUPtr &ProgramUPtr, cl_device_id Device) {
...
+          clEnumVal(compile, "Compile from OpenCL source to .obj"),
```
Comment (dm-vodopyanov) [inline opencl-aot/source/main.cpp:292]:
```suggestion
          clEnumVal(compile, "Compile OpenCL C source to OpenCL program object"),
```
Please re-check it if it's true

File: opencl-aot/source/main.cpp [hunk 10]
```diff
@@ -253,17 +265,46 @@ getCompilerBuildLog(const CLProgramUPtr &ProgramUPtr, cl_device_id Device) {
...
+          clEnumVal(link, "Link OpenCL .obj file(s)")));
```
Comment (dm-vodopyanov) [inline opencl-aot/source/main.cpp:293]:
```suggestion
          clEnumVal(link, "Link OpenCL compiled program objects and creates OpenCL program binary")));
```

File: opencl-aot/source/main.cpp [hunk 11]
```diff
@@ -307,11 +344,37 @@ int main(int Argc, char *Argv[]) {
...
+    OutputFileName = OptOutputElf;
+
+  // Get input files
+  if (!OptInputFile.getNumOccurrences()) {
+    std::cerr << "no input files";
+    return OPENCL_AOT_LIST_OF_INPUT_BINARIES_IS_EMPTY;
+  }
+
+  std::vector<std::string> InputFileNames;
+  // Split input files at ','s and ignore empty file name.
```
Comment (dm-vodopyanov) [inline opencl-aot/source/main.cpp:361]: Multiple input files are passed using space as a delimeter in LLVM-based tools, like
clang source1.cpp source2.cpp
Can you please update `cl::opt<std::string> OptInputFile` (and help message) to support multiple files?

File: opencl-aot/source/main.cpp [hunk 12]
```diff
@@ -420,15 +492,31 @@ int main(int Argc, char *Argv[]) {
...
+      input_programs.push_back(Progs[i].get());
+
+    cl_program Program = clLinkProgram(
+        ContextSPtr.get(), 1, &DeviceId, BuildOptions.c_str(),
+        input_programs.size(), input_programs.data(), nullptr, nullptr, &CLErr);
+    ProgramUPtr.reset(Program);
+  } else {
+    CLErr = clBuildProgram(ProgramUPtr.get(), 1, &DeviceId,
+                           BuildOptions.c_str(), nullptr, nullptr);
+  }
```
Comment (dm-vodopyanov) [inline opencl-aot/source/main.cpp:519]: Line 536
```
std::cerr << formatCLError("Failed to build a program:", CLErr) << '\n';
```
should be customized for compile and link commands

File: opencl-aot/source/main.cpp [hunk 13]
```diff
@@ -56,27 +57,41 @@ auto ContextDeleter = [](cl_context CLCon) {
...
+                          cl_platform_id PlatformId, cl_device_id DeviceId,
+                          CLContextSPtr ContextSPtr) {
   // step 0: define internal types
 
-  enum SupportedTypes : int8_t { BEGIN, ELF = BEGIN, SPIRV, UNKNOWN, END };
+  enum SupportedTypes : int8_t {
+    BEGIN,
+    ELF = BEGIN,
+    SPIRV,
+    LLVMIR,
```
Comment (dm-vodopyanov) [inline opencl-aot/source/main.cpp:84]: Can you add info about LLVM IR to help message?

File: opencl-aot/source/main.cpp [hunk 14]
```diff
@@ -253,17 +265,46 @@ getCompilerBuildLog(const CLProgramUPtr &ProgramUPtr, cl_device_id Device) {
 
 int main(int Argc, char *Argv[]) {
   // step 0: set command line options
-  cl::opt<std::string> OptInputBinary(
-      cl::Positional, cl::Required,
-      cl::desc("<input SPIR-V or OpenCL program binary>"),
+  cl::opt<std::string> OptInputFile(
+      cl::Positional,
+      cl::desc("<input SPIR-V or OpenCL program binary or OpenCL C source>"),
```
Comment (dm-vodopyanov) [inline opencl-aot/source/main.cpp:270]: Not sure if it should be OpenCL program **binary** or OpenCL program **object**

Let's summarize it (please fix me if I was wrong somewhere):

| Input  | Command | Output |
| ------------- | ------------- | ------------- |
| OpenCL C source | build | OpenCL program binary |
| OpenCL C source | compile | OpenCL program object |
| OpenCL C source  | link | **Not valid, show error message** |
| SPIR-V | build | OpenCL program binary |
| SPIR-V | compile | **Not valid, show error message** |
| SPIR-V  | link | OpenCL program binary |
| OpenCL compiled program object | build | OpenCL program binary |
| OpenCL compiled program object | compile | **Not valid, show error message** |
| OpenCL compiled program object | link | OpenCL program binary |
| LLVM IR | build | **The same as for SPIR-V?** |
| LLVM IR | compile | **The same as for SPIR-V?** |
| LLVM IR  | link | **The same as for SPIR-V?** |
| OpenCL program binary | build | **Is it valid use case? Will we get the same program binary?** |
| OpenCL program binary | compile | **Not valid, show error message** |
| OpenCL program binary | link | **Is it valid use case? If it's not valid, show error message** |

File: opencl-aot/source/main.cpp [hunk 15]
```diff
@@ -253,17 +265,46 @@ getCompilerBuildLog(const CLProgramUPtr &ProgramUPtr, cl_device_id Device) {
...
       cl::value_desc("filename"));
+  cl::opt<std::string> OptOutputElf(
+      "o", cl::init("aot-output.bin"),
+      cl::desc("Specify the output OpenCL program binary filename"),
```
Comment (dm-vodopyanov) [inline opencl-aot/source/main.cpp:274]: Am I correct that during --cmd=compile, we will get OpenCL program object, not binary?

File: opencl-aot/source/main.cpp [hunk 16]
```diff
@@ -484,7 +572,7 @@ int main(int Argc, char *Argv[]) {
     return OPENCL_AOT_FAILED_TO_CREATE_ELF;
   }
   std::cout << "OpenCL program binary file was successfully created: "
```
Comment (dm-vodopyanov) [inline opencl-aot/source/main.cpp:574]: If during --cmd=compile, we will get OpenCL program object, this message should be changed

**Review by dm-vodopyanov — CHANGES_REQUESTED** [review]
Very sorry for delay in review.

File: opencl-aot/source/main.cpp [hunk 17]
```diff
@@ -253,17 +265,46 @@ getCompilerBuildLog(const CLProgramUPtr &ProgramUPtr, cl_device_id Device) {
 
 int main(int Argc, char *Argv[]) {
   // step 0: set command line options
-  cl::opt<std::string> OptInputBinary(
-      cl::Positional, cl::Required,
-      cl::desc("<input SPIR-V or OpenCL program binary>"),
+  cl::opt<std::string> OptInputFile(
+      cl::Positional,
```
[the remark this answers is not in the document]
Comment (AlexeySachkov) [reply opencl-aot/source/main.cpp:269]: If I understand correctly, we now have a few aliases to that option: I don't know how exactly `cl::opt` works here, but we need to ensure that the tool doesn't require from us an input file if it was passed through `-input`/`-spv` or some other new option

File: opencl-aot/source/main.cpp (hunk 14 above)
Comment (AlexeySachkov) [reply opencl-aot/source/main.cpp:270]: > SPIR-V | compile | Not valid, show error message
> -- | -- | --

This is a valid call of OpenCL API. You can "compile" a SPIR-V file and then use link API to merge several SPIR-V files together

File: opencl-aot/source/main.cpp [hunk 18]
```diff
@@ -253,17 +265,46 @@ getCompilerBuildLog(const CLProgramUPtr &ProgramUPtr, cl_device_id Device) {
...
+      cl::values(
+          clEnumVal(
+              build,
+              "Build executable IR from OpenCL source or OpenCL .obj file"),
+          clEnumVal(compile, "Compile from OpenCL source to .obj"),
+          clEnumVal(link, "Link OpenCL .obj file(s)")));
+
+  // These options are only needed for compatibility with old tool.
+  // TODO: Delete these options once transition to a regular -o and regular
+  // input argument is complete.
```
[the remark this answers is not in the document]
Comment (cdai2) [reply opencl-aot/source/main.cpp:297]: aoc tool will be replaced by this tool.  This is why link add these changes.

[the remark this answers is not in the document]
Comment (dm-vodopyanov) [reply opencl-aot/source/main.cpp:297]: I thought there will be some intermediate step, in which aoc calls opencl-aot instead of aocl-ioc64.
In this case, it is even easier - as separate patch, Clang driver should be updated somewhere [here](https://github.com/intel/llvm/blob/sycl/clang/lib/Driver/ToolChains/SYCL.cpp#L304) in the same way as it was done for opencl-aot integration for x86 [here](https://github.com/intel/llvm/blob/sycl/clang/lib/Driver/ToolChains/SYCL.cpp#L462).

File: opencl-aot/source/main.cpp (hunk 14 above)
Comment (linkren) [reply opencl-aot/source/main.cpp:270]: > OpenCL compiled program object | compile | Not valid, show error message
> -- | -- | --

This is also valid. The output of compiling a OpenCL object file is itself.

File: opencl-aot/source/main.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (linkren) [reply opencl-aot/source/main.cpp:297]: I'm not familiar with Clang driver. My suggestion is to implement these deprecated options in intel customizations style if community doesn't accept.

[the remark this answers is not in the document]
Comment (AlexeySachkov) [reply opencl-aot/source/main.cpp:297]: We need to understand all the use-cases here: in intel/llvm we indeed could avoid updating `opencl-aot` tool and just update the driver to use it instead of `aocl-ioc64`.

However, if this tool is used in some other workflow/toolchain somewhere (like a downstream project) and we are not ready to change that workflow to pass new options yet, then this patch is indeed needed. But please note that we will do extra work twice: in this PR we introducing compatibility option and later we will have another PR which removes them. If we actually could update our usage model of the tool, it would be better (1 task instead of 3)

File: opencl-aot/source/main.cpp [hunk 19]
```diff
@@ -56,27 +57,41 @@ auto ContextDeleter = [](cl_context CLCon) {
...
+    UNKNOWN,
+    END
+  };
   static std::map<SupportedTypes, const std::string> SupportedTypesToNames{
-      {ELF, "OpenCL program binary"}, {SPIRV, "SPIR-V"}, {UNKNOWN, "UNKNOWN"}};
+      {ELF, "OpenCL program binary"},
+      {SPIRV, "SPIR-V"},
+      {LLVMIR, "LLVM IR bitcode"},
+      {SOURCE, "OpenCL source"},
+      {UNKNOWN, "UNKNOWN"}};
```
[the remark this answers is not in the document]
Comment (linkren) [reply opencl-aot/source/main.cpp:94]: I finally didn't remove it because I implement a function to identify .cl files.

File: opencl-aot/source/main.cpp [hunk 20]
```diff
@@ -420,15 +492,31 @@ int main(int Argc, char *Argv[]) {
...
-  // step 8: build OpenCL program
-  CLErr = clBuildProgram(ProgramUPtr.get(), 1, &DeviceId, BuildOptions.c_str(),
-                         nullptr, nullptr);
+  // step 8: compile | build | link OpenCL program
+  if (OptCommand.getNumOccurrences() && OptCommand == Commands::compile) {
+    CLErr =
+        clCompileProgram(ProgramUPtr.get(), 1, &DeviceId, BuildOptions.c_str(),
+                         0, nullptr, nullptr, nullptr, nullptr);
+  } else if (OptCommand.getNumOccurrences() && OptCommand == Commands::link &&
+             Progs.size() > 1) {
```
[the remark this answers is not in the document]
Comment (linkren) [reply opencl-aot/source/main.cpp]: If Progs.size() == 1, link is equivalent to build. And Progs can't be empty.

File: opencl-aot/source/main.cpp (hunk 13 above)
Comment (linkren) [reply opencl-aot/source/main.cpp:84]: The LLVM IR objects are mostly like OpenCL compiled object files.

**Comment by bader** [comment]
@AlexeySachkov, @dm-vodopyanov, ping.

File: opencl-aot/source/main.cpp (hunk 14 above)
Comment (dm-vodopyanov) [reply opencl-aot/source/main.cpp:276]: Thanks @AlexeySachkov and @linkren. @linkren , can you please include missing workflows to the source code?

Comment (linkren) [reply opencl-aot/source/main.cpp:276]: They are already there. There is no specific error message but internal returned error is displayed for invalid workflows. Users should know the type of input file they feed to this tool.

File: opencl-aot/source/utils.cpp [hunk 21]
```diff
@@ -412,6 +417,10 @@ bool isFileStartsWithGivenMagicNumber(const std::vector<char> &BinaryData,
   return BinaryDataAsIntBuffer[0] == ExpectedMagicNumber;
 }
 
+bool isFileOCLSource(const std::string &FileName) {
+  return isFileEndsWithGivenExtentionName(FileName, ".cl");
```
[the remark this answers is not in the document]
Comment (linkren) [reply opencl-aot/source/utils.cpp:421]: Sure. I asked @cdai2. We will only accept .cl files and .h is expected to be passed to -I in --bo option.

File: opencl-aot/source/main.cpp (hunk 14 above)
Comment (linkren) [reply opencl-aot/source/main.cpp:276]: They are already there. There is no specific error message but internal returned error is displayed for invalid workflows. Users should know the type of input file they feed to this tool.

File: opencl-aot/source/utils.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (cdai2) [reply opencl-aot/source/utils.cpp:421]: Agree with @linkren.

File: opencl-aot/source/main.cpp (hunk 13 above)
Comment (linkren) [reply opencl-aot/source/main.cpp:84]: Done.

**Review by dm-vodopyanov — APPROVED** [review]
